### PR TITLE
Cache in term_of_fconstr

### PR DESCRIPTION
**Pull request by SkySkimmer**
Depends:
- https://github.com/coq/coq/pull/19837

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Seems pretty mediocre

**Comment by Janno** [comment]
Just for the record, here's the result on BlueRock's code.


| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   +0.27% |  35982.3 |  36078.9 |    +96.7 | cpp2v-generated
|   -1.43% |  88256.6 |  86994.5 |  -1262.1 | other
|   -0.94% | 124238.9 | 123073.5 |  -1165.4 | total

 I see that this is not a strict win across the board for all developments so I understand why it was closed. But I wonder if it could be possible to combine this with https://github.com/coq/coq/pull/19449 in some way. For terms being converted from `constr` to `fconstr`, instead of starting with `None`, we'd remember the original `constr` with a tag to make sure we know to discard it once we made any changes to the `fconstr`. I am not sure how easy it is to detect when to discard the term but for the cases covered by #19449 they should never need to be discarded since those are ground terms.

File: kernel/cClosure.ml [hunk 1]
```diff
@@ -260,7 +266,7 @@ let zupdate info m s =
   let share = info.i_cache.i_share in
   if share && is_red m.mark then
     let s' = compact_stack m s in
-    let _ = m.term <- FLOCKED in
+    let _ = m.term <- FLOCKED; m.as_constr <- None in
```
Comment (SkySkimmer) [inline kernel/cClosure.ml:269]: @Janno I guess this is what you mean by "discard the term when it's changed"

Comment (Janno) [reply kernel/cClosure.ml:269]: My understanding of the mutable parts of CClosure is still very bad. This certainly seems like one important point where any cached `constr` value would need to be discarded. I was under the impression that there were places were `fconstr`s were constructed from existing `fconstr`s but that does not seem to be the case. So maybe this is indeed the only place that needs to reset to `None`.

File: kernel/cClosure.ml [hunk 2]
```diff
@@ -301,33 +307,34 @@ let destFLambda clos_fun t =
   | FLambda(_,[(na,ty)],b,e) ->
     (usubst_binder e na,clos_fun e ty,clos_fun (usubs_lift e) b)
   | FLambda(n,(na,ty)::tys,b,e) ->
-    (usubst_binder e na,clos_fun e ty,{mark=t.mark;term=FLambda(n-1,tys,b,usubs_lift e)})
+    (usubst_binder e na,clos_fun e ty,{mark=t.mark;term=FLambda(n-1,tys,b,usubs_lift e); as_constr=None})
   | _ -> assert false
 
 (* Optimization: do not enclose variables in a closure.
    Makes variable access much faster *)
+(* XXX some of these as_constr could be "Some t" *)
```
Comment (Janno) [inline kernel/cClosure.ml:315]: I think my vague proposal boils down to the same thing as this comment here, in `mk_clos` and likely in all other places where `FInd`, `FConstruct`, `FFlex` (maybe not all cases of that?), `FAtom`, and primitive nodes such as `FFloat`, `FInt`, and `FString` are turned into `fconstr`. We should be able to get away with `Some (ref (Closed <original term>))` for most of these, right?

**Pull request closed without merging**